File: src/synthorg/api/controllers/_bulk_delete.py
# module-kind: code
"""One operator action, one request: deleting a selected set.

A dashboard bulk delete was a loop of single-row DELETEs, and every one of the
three destructive operations it drives is rate limited per user (five projects a
minute, five plans, twenty tasks). Selecting more than that had the limiter
refuse the tail, so the action reported a partial failure for a reason that had
nothing to do with the rows: an operator clearing a round's residue got five
deletions and a wall of 5001s.

The limiter is not the thing to loosen. It bounds destructive throughput, and a
bulk delete is ONE destructive decision the operator took once, so it is one
call with its own budget rather than N calls against a per-row one.

Every row is still deleted by the same code path its own DELETE takes, so the
cascade, the tombstone and the WebSocket event are identical; what differs is
that a refusal is collected rather than raised, because one row that cannot go
must not decide the fate of the others.
"""

import asyncio
from collections.abc import Awaitable, Callable
from typing import Final

from pydantic import BaseModel, ConfigDict, Field

from synthorg.api.state import AppState
from synthorg.core.clock import Clock
from synthorg.core.critical_errors import reraise_critical
from synthorg.core.domain_errors import DomainError
from synthorg.core.types import NotBlankStr
from synthorg.observability import get_logger, safe_error_description
from synthorg.observability.events.api import (
    API_BULK_DELETE_PARTIAL,
    API_BULK_DELETE_ROW_REFUSED,
)
from synthorg.settings.enums import SettingNamespace
from synthorg.settings.state import SettingsStateSlice, config_resolver_of

logger = get_logger(__name__)

#: How many rows one call may carry, and the real bound on how much one
#: request can destroy: the rate limit meters requests, so the per-user ceiling
#: is this number times that budget rather than the budget alone.
#:
#: Held at the largest page the dashboard can render, because selection is
#: scoped to the rows on screen and no operator action can produce a bigger
#: set. A larger selection is two actions, which the operator can see and stop
#: between.
MAX_BULK_DELETE_IDS: Final[int] = 100

#: What a row that was never attempted is told. Worded for the operator and
#: distinct from a refusal the row itself caused, because the remedy differs:
#: nothing is wrong with these rows and selecting them again is the whole fix.
_BUDGET_SPENT_REASON: Final[str] = (
    "not attempted: the request reached its time budget before this row; "
    "select it again to finish"
)


class BulkDeleteRequest(BaseModel):
    """The rows an operator selected for deletion.

    Attributes:
        ids: The identifiers to delete, in any order. Duplicates are collapsed.
    """

    model_config = ConfigDict(frozen=True, allow_inf_nan=False, extra="forbid")

    ids: tuple[NotBlankStr, ...] = Field(
        min_length=1,
        max_length=MAX_BULK_DELETE_IDS,
        description="Identifiers of the rows to delete.",
    )


class BulkDeleteFailure(BaseModel):
    """One row that could not be deleted, and why.

    Attributes:
        id: The row the operator selected. It is echoed rather than resolved to
            a name because the caller selected it from a list it is holding, so
            it labels its own row; nothing renders this value.
        reason: The refusing error's own operator-facing message. Deliberately
            not the log description: that one preserves the exception class
            name and whatever identifier the raise interpolated, and this field
            is rendered in a toast, where an internal class name and a database
            key are both things no operator surface may print. The specific
            cause is kept on ``API_BULK_DELETE_ROW_REFUSED`` instead.
    """

    model_config = ConfigDict(frozen=True, allow_inf_nan=False, extra="forbid")

    id: NotBlankStr = Field(description="Identifier of the row that remains.")
    reason: NotBlankStr = Field(description="Why this row could not be deleted.")


class BulkDeleteResult(BaseModel):
    """What one bulk delete actually did.

    Both halves are reported, because a partial outcome is neither a success
    nor a failure and telling an operator only about the failures is how they
    conclude that nothing happened to an action they cannot undo.

    Attributes:
        deleted: Identifiers that were removed.
        failed: The rows that remain, each naming what refused.
    """

    model_config = ConfigDict(frozen=True, allow_inf_nan=False, extra="forbid")

    deleted: tuple[NotBlankStr, ...] = Field(
        default=(),
        description="Identifiers that were removed.",
    )
    failed: tuple[BulkDeleteFailure, ...] = Field(
        default=(),
        description="Rows that remain, each with the reason it does.",
    )


async def resolve_bulk_delete_budget(app_state: AppState) -> float:
    """Resolve the per-request bulk-delete budget, live.

    Reads ``api.bulk_delete_budget_seconds`` through the settings chain (DB >
    env > default) so a change applies without a restart. Resolver-read-only,
    so the boot-config value is the correct fallback when the resolver is
    absent or unwell: a settings hiccup must not remove the bound on a
    destructive request.

    Args:
        app_state: Application state.

    Returns:
        The ceiling in seconds.

    Raises:
        CancelledError: Propagated when the resolver await is cancelled.
    """
    boot_value = app_state.config.api.bulk_delete_budget_seconds
    if app_state.slice(SettingsStateSlice).config_resolver is None:
        return boot_value
    try:
        return await config_resolver_of(app_state).get_float(
            SettingNamespace.API.value,
            "bulk_delete_budget_seconds",
        )
    except asyncio.CancelledError:
        raise
    except Exception as exc:  # noqa: BLE001 -- criticals re-raised
        reraise_critical(exc)
        logger.warning(
            API_BULK_DELETE_PARTIAL,
            setting="api.bulk_delete_budget_seconds",
            error_type=type(exc).__name__,
            error=safe_error_description(exc),
            fallback_seconds=boot_value,
        )
        return boot_value


async def run_bulk_delete(
    ids: tuple[NotBlankStr, ...],
    delete_one: Callable[[str], Awaitable[None]],
    *,
    entity: str,
    clock: Clock,
    budget_seconds: float,
) -> BulkDeleteResult:
    """Delete each of *ids* through *delete_one*, collecting refusals.

    Sequential rather than concurrent: each of these cascades across plans,
    tasks and approvals, and running them together on one operator click would
    interleave writes over rows that reference each other.

    Sequential also means the request grows with the selection, and the caller
    waiting on it gives up on its own schedule. Past *budget_seconds* the loop
    stops before starting another row and reports the rest as refused, so the
    answer arrives while the browser is still listening and covers every row
    the operator selected. Without it the deletions still happen and the
    operator is told they did not, which is worse than not starting them.

    Args:
        ids: The selected identifiers; duplicates are collapsed, order kept.
        delete_one: The same removal the row's own DELETE performs.
        clock: Time source, injected so the budget is testable.
        budget_seconds: Wall-clock ceiling for the whole call.
        entity: What is being deleted, for the log.

    Returns:
        What was removed and what remains.
    """
    seen: set[str] = set()
    deleted: list[NotBlankStr] = []
    failed: list[BulkDeleteFailure] = []
    started_at = clock.now()
    for entity_id in ids:
        if entity_id in seen:
            continue
        seen.add(entity_id)
        # Checked BEFORE the row rather than after: the budget exists to keep
        # the reply inside the caller's wait, and a check afterwards has
        # already spent the time it was meant to protect.
        if (clock.now() - started_at).total_seconds() >= budget_seconds:
            failed.append(
                BulkDeleteFailure(
                    id=NotBlankStr(entity_id),
                    reason=NotBlankStr(_BUDGET_SPENT_REASON),
                )
            )
            continue
        try:
            await delete_one(entity_id)
        except DomainError as exc:
            # A typed refusal is an outcome for THIS row (still building, no
            # longer there, an approval decided mid-flight). Anything else is
            # not about the row and takes the request down as it would a
            # single delete.
            logger.warning(
                API_BULK_DELETE_ROW_REFUSED,
                entity=entity,
                error_type=type(exc).__name__,
                error=safe_error_description(exc),
            )
            failed.append(
                BulkDeleteFailure(
                    id=NotBlankStr(entity_id),
                    reason=NotBlankStr(exc.default_message),
                )
            )
            continue
        deleted.append(NotBlankStr(entity_id))
    if failed:
        logger.warning(
            API_BULK_DELETE_PARTIAL,
            entity=entity,
            deleted_count=len(deleted),
            failed_count=len(failed),
        )
    return BulkDeleteResult(deleted=tuple(deleted), failed=tuple(failed))


__all__ = [
    "MAX_BULK_DELETE_IDS",
    "BulkDeleteFailure",
    "BulkDeleteRequest",
    "BulkDeleteResult",
    "run_bulk_delete",
]

File: web/src/stores/_bulk-delete.ts
/**
 * One call, one toast, for every list that deletes a selection.
 *
 * Each of these endpoints is rate limited per user, so a client loop over N
 * rows refused its own tail for a reason that had nothing to do with the rows:
 * the backend takes the whole selection in one request and answers per row.
 *
 * The toast lives here rather than in each store because a partial outcome is
 * neither a success nor a failure, and three copies of that decision is how one
 * list comes to report "failed" for an action that half worked.
 */

import { createLogger } from '@/lib/logger'
import { useToastStore } from '@/stores/toast'
import { formatBatchErrors, getCrudErrorTitle, getErrorMessage } from '@/utils/errors'
import { sanitizeForLog } from '@/utils/logging'
import type { BulkDeleteResult } from '@/api/types/bulk-delete'

const log = createLogger('bulk-delete')

/** What the caller reports back to its page. */
export interface BulkDeleteOutcome {
  succeeded: number
  failed: number
  failedReasons: string[]
  /**
   * The rows the backend confirmed it removed.
   *
   * Carried as ids rather than a count because the selection that survives a
   * partial delete is the rows still on screen, and a count cannot say which
   * those are.
   */
  deletedIds: readonly string[]
}

/** The noun this list deletes, singular and plural. */
export interface BulkDeleteNoun {
  readonly one: string
  readonly many: string
}

export interface BulkDeleteArgs {
  readonly ids: readonly string[]
  /** The single request that does the deleting. */
  readonly call: (ids: readonly string[]) => Promise<BulkDeleteResult>
  /** Drop the rows the backend removed from the list on screen. */
  readonly removeRows: (ids: readonly string[]) => void
  readonly noun: BulkDeleteNoun
}

function emitToast(
  outcome: BulkDeleteOutcome,
  total: number,
  noun: BulkDeleteNoun,
): void {
  const description = outcome.failedReasons.length > 0
    ? formatBatchErrors(outcome.failedReasons)
    : undefined
  if (outcome.failed === 0 && outcome.succeeded > 0) {
    useToastStore.getState().add({
      variant: 'success',
      title: outcome.succeeded === 1
        ? `${noun.one} deleted`
        : `${String(outcome.succeeded)} ${noun.many} deleted`,
    })
    return
  }
  if (outcome.succeeded > 0) {
    useToastStore.getState().add({
      variant: 'warning',
      title: `Deleted ${String(outcome.succeeded)} of ${String(total)} ${noun.many}`,
      description,
    })
    return
  }
  emitTotalFailureToast(outcome, total, noun, description)
}

function emitTotalFailureToast(
  outcome: BulkDeleteOutcome,
  total: number,
  noun: BulkDeleteNoun,
  description: string | undefined,
): void {
  // Nothing was deleted, so the whole selection is what was lost. Counting
  // only the rows the backend named would render "Failed to delete 0" for an
  // answer that reported neither a deletion nor a refusal.
  const lost = outcome.failed === 0 ? total : outcome.failed
  useToastStore.getState().add({
    variant: 'error',
    ...getCrudErrorTitle(
      new Error(outcome.failedReasons[0] ?? `Failed to delete ${noun.many}`),
      lost === 1
        ? `Failed to delete ${noun.one.toLowerCase()}`
        : `Failed to delete ${String(lost)} ${noun.many}`,
    ),
    description,
  })
}

/**
 * Delete *ids* in one request and report what happened.
 *
 * Rows are removed AFTER the answer rather than optimistically: the backend
 * says which ones went, so there is nothing to guess at and nothing to roll
 * back. Returns `false` when the request itself failed, which is the sentinel
 * the store-mutation contract uses for a call that settled nothing.
 */
export async function runBulkDelete({
  ids,
  call,
  removeRows,
  noun,
}: BulkDeleteArgs): Promise<BulkDeleteOutcome | false> {
  const uniqueIds = Array.from(new Set(ids))
  // An empty selection is settled here rather than at the backend, which
  // refuses it: sending it would spend a request against a per-user rate limit
  // and answer an operator who selected nothing with a failure toast.
  if (uniqueIds.length === 0) {
    return false
  }
  // The guard covers the whole body, not just the call. Everything after it is
  // caller-supplied (the row removal each store passes in) or another store's
  // (the toast), so a throw there would escape a store mutation, which the
  // contract forbids, and strand the confirm dialog its caller cannot close
  // while a delete is in flight.
  try {
    const result: BulkDeleteResult = await call(uniqueIds)
    removeRows(result.deleted)
    const outcome: BulkDeleteOutcome = {
      succeeded: result.deleted.length,
      failed: result.failed.length,
      failedReasons: result.failed.map((failure) => failure.reason),
      deletedIds: result.deleted,
    }
    if (outcome.failed > 0) {
      log.error(`Bulk delete ${noun.many} partial`, sanitizeForLog({
        failed: outcome.failed,
        reasons: outcome.failedReasons,
      }))
    }
    emitToast(outcome, uniqueIds.length, noun)
    // The sentinel means the call settled nothing, so it keys on what was
    // deleted rather than on what refused: an answer naming no deletion and no
    // refusal settled nothing either, and reporting that as success tells the
    // caller to clear a selection still sitting in the list.
    return outcome.succeeded === 0 ? false : outcome
  } catch (err) {
    log.error(`Bulk delete ${noun.many} failed`, sanitizeForLog(err))
    useToastStore.getState().add({
      variant: 'error',
      ...getCrudErrorTitle(err, `Failed to delete ${noun.many}`),
      description: getErrorMessage(err),
    })
    return false
  }
}
